test/framework/clusterctl/clusterctl_helpers.go

```diff
@@ -274,42 +282,67 @@ func ApplyClusterTemplateAndWait(ctx context.Context, input ApplyClusterTemplate
 	result.Cluster = &clusterv1.Cluster{
 		ObjectMeta: metav1.ObjectMeta{
-			Name:      input.ConfigCluster.ClusterName,
-			Namespace: input.ConfigCluster.Namespace,
+			Name:      clusterName,
+			Namespace: namespace,
 		},
 	}
 
-	log.Logf("Getting the cluster template yaml")
-	workloadClusterTemplate := ConfigCluster(ctx, ConfigClusterInput{
-		// pass reference to the management cluster hosting this test
-		KubeconfigPath: input.ConfigCluster.KubeconfigPath,
-		// pass the clusterctl config file that points to the local provider repository created for this test,
-		ClusterctlConfigPath: input.ConfigCluster.ClusterctlConfigPath,
-		// select template
-		Flavor: input.ConfigCluster.Flavor,
-		// define template variables
-		Namespace:                input.ConfigCluster.Namespace,
-		ClusterName:              input.ConfigCluster.ClusterName,
-		KubernetesVersion:        input.ConfigCluster.KubernetesVersion,
-		ControlPlaneMachineCount: input.ConfigCluster.ControlPlaneMachineCount,
-		WorkerMachineCount:       input.ConfigCluster.WorkerMachineCount,
-		InfrastructureProvider:   input.ConfigCluster.InfrastructureProvider,
-		// setup clusterctl logs folder
-		LogFolder:           input.ConfigCluster.LogFolder,
-		ClusterctlVariables: input.ConfigCluster.ClusterctlVariables,
-	})
-	Expect(workloadClusterTemplate).ToNot(BeNil(), "Failed to get the cluster template")
+	var workloadClusterTemplate []byte
+
+	if input.WorkloadClusterTemplate != nil {
+		log.Logf("Using the provided cluster template yaml")
+		workloadClusterTemplate = input.WorkloadClusterTemplate.YAML
+	} else {
+		log.Logf("Getting the cluster template yaml")
+		workloadClusterTemplate = ConfigCluster(ctx, ConfigClusterInput{
+			// pass reference to the management cluster hosting this test
+			KubeconfigPath: input.ConfigCluster.KubeconfigPath,
+			// pass the clusterctl config file that points to the local provider repository created for this test,
+			ClusterctlConfigPath: input.ConfigCluster.ClusterctlConfigPath,
+			// select template
+			Flavor: input.ConfigCluster.Flavor,
+			// define template variables
+			Namespace:                input.ConfigCluster.Namespace,
+			ClusterName:              input.ConfigCluster.ClusterName,
+			KubernetesVersion:        input.ConfigCluster.KubernetesVersion,
+			ControlPlaneMachineCount: input.ConfigCluster.ControlPlaneMachineCount,
+			WorkerMachineCount:       input.ConfigCluster.WorkerMachineCount,
+			InfrastructureProvider:   input.ConfigCluster.InfrastructureProvider,
+			// setup clusterctl logs folder
+			LogFolder:           input.ConfigCluster.LogFolder,
+			ClusterctlVariables: input.ConfigCluster.ClusterctlVariables,
+		})
+		Expect(workloadClusterTemplate).ToNot(BeNil(), "Failed to get the cluster template")
+	}
 
-	log.Logf("Applying the cluster template yaml to the cluster")
+	log.Logf("Applying the cluster template yaml of cluster %s to the cluster", klog.KRef(namespace, clusterName))
```

Comment:
```suggestion
	log.Logf("Applying the cluster template yaml of cluster %s", klog.KRef(namespace, clusterName))
```
nit: redundant and a bit confusing